### PR TITLE
Fix handling of 1d table elements with single entry in fits

### LINKED ISSUE #10765: Table reading/writing to/from fits does not support writing 1d arrays with a single element
### Description

Astropy `Table` writing and reading to/from fits looses the information about dimensionality,
if the cell data written is a 1d array with a single entry.
Then, on reading back the data, it will be read as a number, so ndim=0, instead of ndim=1.

### Expected behavior

Preserve this informatino

### Actual behavior

Wrong dimensionality

### Steps to Reproduce

```python

❯ python
Python 3.8.5 (default, Sep  5 2020, 10:50:12) 
[GCC 10.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from astropy.table import Table
>>> t = Table({'x': [[1], [2]]})
>>> t.write('/tmp/test.fits', overwrite=True)
>>> t['x'].shape
(2, 1)
>>> Table.read('/tmp/test.fits')['x'].shape
(2,)

```

The error seems to be in both reading and writing.

The writing part fails to add the `TDIM` keyword and the reading part has a regex that does not match a tuple with just one entry.

### System Details

```
Linux-5.8.10-1-MANJARO-x86_64-with-glibc2.2.5
Python 3.8.5 (default, Sep  5 2020, 10:50:12) 
[GCC 10.2.0]
Numpy 1.19.2
astropy 4.0.1.post1
Scipy 1.5.2
Matplotlib 3.3.2
```

### PR DESCRIPTION
**Pull request by MaxNoe**
### Description

Add test for and fix writing and reading of fits columns that contain single values in 1d columns
This pull request is to address #10765 

Fixes #10765

### CODE REVIEW

**Comment by MaxNoe** [comment]
regex101 links for the new and the old regex:

new: https://regex101.com/r/jlQzyT/2
old: https://regex101.com/r/YseaAK/1

File: CHANGES.rst [hunk 1]
```diff
@@ -47,6 +47,8 @@ astropy.io.ascii
 astropy.io.fits
 ^^^^^^^^^^^^^^^
 
+- Fix handling of 1-dimensional arrays with single elements in BinaryTableHDUs [#10768]
```
Comment (dhomeier) [inline CHANGES.rst:50]:
```suggestion
- Fix handling of 1-dimensional arrays with single elements in ``BinTableHDU``s [#10768]
```
Suggest to use the class name; otherwise looks good once it's moved to the new milestone.

**Review by dhomeier — CHANGES_REQUESTED** [review]

File: CHANGES.rst (hunk 1 above)
Comment (MaxNoe) [reply CHANGES.rst]: Both done

**Comment by MaxNoe** [comment]
@dhomeier The build seems to fail due to an unrelated problem to my changes. Any idea what happened there?

**Review by dhomeier — APPROVED** [review]

**Comment by dhomeier** [comment]
> @dhomeier The build seems to fail due to an unrelated problem to my changes. Any idea what happened there?

This appears to be caused by a simultaneous pytest update on both CI platforms, just reported in #10773.

Otherwise looks good to me; with only changelog changes could probably even have gone in with a `[ci skip]`!

**Comment by MaxNoe** [comment]
> Otherwise looks good to me; with only changelog changes could probably even have gone in with a [ci skip]!

Yes, sorry forgot about that

**Comment by MaxNoe** [comment]
Thanks for getting this into the next  point release

File: astropy/io/fits/tests/test_table.py [hunk 2]
```diff
@@ -545,7 +545,7 @@ def test_appending_a_column(self):
         for i in range(len(t1[1].columns)):
             hdu.data.field(i)[nrows1:] = t2[1].data.field(i)
 
-        hdu.writeto(self.temp('newtable.fits'))
+        hdu.writeto(self.temp('newtable.fits'), overwrite=True,)
```
Comment (saimn) [inline astropy/io/fits/tests/test_table.py:548]: I guess you needed this for debugging, but could you remove it now ?

File: astropy/io/fits/tests/test_table.py [hunk 3]
```diff
@@ -1702,12 +1702,14 @@ def test_multi_dimensional_columns(self):
...
-        thdu.header['TDIM1'] = '(2,3)'
-        thdu.header['TDIM2'] = '(4,2)'
 
         thdu.writeto(self.temp('newtable.fits'))
 
+        with fits.open(self.temp('newtable.fits'), mode='update') as hdul:
+            # Modify the TDIM fields to my own specification
+            hdul[1].header['TDIM1'] = '(2,3)'
+            hdul[1].header['TDIM2'] = '(4,2)'
+
```
Comment (saimn) [inline astropy/io/fits/tests/test_table.py:1712]: Why is this change needed ?

File: astropy/io/fits/tests/test_table.py [hunk 4]
```diff
@@ -1753,6 +1755,19 @@ def test_multi_dimensional_columns(self):
...
+        thdu = fits.BinTableHDU.from_columns(data)
+
+        thdu.writeto(self.temp('onedtable.fits'))
+
+        with fits.open(self.temp('onedtable.fits')) as hdul:
+            thdu = hdul[1]
+
+            c = thdu.data.field(0)
+            assert c.shape == (2, 1)
+
```
Comment (saimn) [inline astropy/io/fits/tests/test_table.py:1770]: Could you also add a check for `TDIM` ?

Comment (saimn) [reply astropy/io/fits/tests/test_table.py:1770]: Also it could be useful to add a test with `astropy.table.Table` (in `test_connect.py`) to make sure the use case of your original issue is fixed.

**Review by saimn — CHANGES_REQUESTED** [review]
Thanks for the fix @MaxNoe, and for the regex101 links ;)
Just a few minor comments, otherwise looks good !

File: astropy/io/fits/tests/test_table.py (hunk 2 above)
Comment (MaxNoe) [reply astropy/io/fits/tests/test_table.py:548]: yes

File: astropy/io/fits/tests/test_table.py (hunk 3 above)
Comment (MaxNoe) [reply astropy/io/fits/tests/test_table.py:1712]: At least when running the tests locally, it seemed that setting these headers *before* writing the file was overwritten by the actual dimensions of the data.

So I had to change to modify them after the file was initially written.

Comment (MaxNoe) [reply astropy/io/fits/tests/test_table.py:1712]: This did not trigger before, as TDIM was only added `if len(shape) > 1`, which is no longer the case.

File: astropy/io/fits/tests/test_table.py (hunk 4 above)
Comment (MaxNoe) [reply astropy/io/fits/tests/test_table.py:1786]: done

**Comment by MaxNoe** [comment]
@saimn I think I addressed all your comments, please have a look again

**Comment by MaxNoe** [comment]
@dhomeier The tests are passing again

**Comment by saimn** [comment]
Looks good now, thanks @MaxNoe !
